# Homework: Sonar signal classification
In this workshop you will train a binary classification neural network to differentiate if sonar signals bounced against a metal cylinder or a rock.

[Info of the data](https://archive.ics.uci.edu/ml/datasets/Connectionist+Bench+(Sonar,+Mines+vs.+Rocks))

Goal: Get an accuracy higher than 0.80 in the validation set

## Get the data from Google Drive

In [50]:
# Import libraries to interact with Google Drive
from pydrive.auth import GoogleAuth
from pydrive.drive import GoogleDrive
from google.colab import auth
from oauth2client.client import GoogleCredentials

In [51]:
# Authenticate with your Google account to get access to the data
auth.authenticate_user()
gauth = GoogleAuth()
gauth.credentials = GoogleCredentials.get_application_default()
drive = GoogleDrive(gauth)

In [52]:
# Download data
download = drive.CreateFile({'id': '1rw5l3jCo2vlNc8NLrMk3KPZr6HsjNRCY'})
download.GetContentFile('sonar.csv')

In [53]:
ls

sample_data/  sonar.csv


## Do some magic below! ;)

In [54]:
pip install ucimlrepo

In [55]:
from ucimlrepo import fetch_ucirepo

# fetch dataset
connectionist_bench_sonar_mines_vs_rocks = fetch_ucirepo(id=151)

# data (as pandas dataframes)
X = connectionist_bench_sonar_mines_vs_rocks.data.features
y = connectionist_bench_sonar_mines_vs_rocks.data.targets

# metadata
print(connectionist_bench_sonar_mines_vs_rocks.metadata)

# variable information
print(connectionist_bench_sonar_mines_vs_rocks.variables)

{'uci_id': 151, 'name': 'Connectionist Bench (Sonar, Mines vs. Rocks)', 'repository_url': 'https://archive.ics.uci.edu/dataset/151/connectionist+bench+sonar+mines+vs+rocks', 'data_url': 'https://archive.ics.uci.edu/static/public/151/data.csv', 'abstract': 'The task is to train a network to discriminate between sonar signals bounced off a metal cylinder and those bounced off a roughly cylindrical rock.', 'area': 'Physics and Chemistry', 'tasks': ['Classification'], 'characteristics': ['Multivariate'], 'num_instances': 208, 'num_features': 60, 'feature_types': ['Real'], 'demographics': [], 'target_col': ['class'], 'index_col': None, 'has_missing_values': 'no', 'missing_values_symbol': None, 'year_of_dataset_creation': 1988, 'last_updated': None, 'dataset_doi': '10.24432/C5T01Q', 'creators': ['Terry Sejnowski', 'R. Gorman'], 'intro_paper': None, 'additional_info': {'summary': 'The file "sonar.mines" contains 111 patterns obtained by bouncing sonar signals off a metal cylinder at various a

In [56]:
import pandas as pd

data = pd.read_csv("sonar.csv", header=None)

#Divide los datos en características y etiquetas
X = data.iloc[:, :-1].values
y = data.iloc[:, -1].map({'R': 0, 'M': 1}).values

In [57]:
from sklearn.model_selection import train_test_split

X_train, X_val, y_train, y_val = train_test_split(X, y, test_size=0.2, random_state=40)
#20% de datos para validación

In [58]:
from sklearn.ensemble import RandomForestClassifier

classifier = RandomForestClassifier(random_state=40)
classifier.fit(X_train, y_train)

RandomForestClassifier(random_state=40)

In [59]:
from sklearn.metrics import accuracy_score, classification_report

#Aquí hace predicciones en el conjunto de validación
y_pred = classifier.predict(X_val)

accuracy = accuracy_score(y_val, y_pred)
print("Precisión de", accuracy)

print(classification_report(y_val, y_pred))
#Lo que hace el modelo es asociar los valores de cada columna o característica con la etiqueta correspondiente.

Precisión de 0.8809523809523809
              precision    recall  f1-score   support

           0       0.93      0.76      0.84        17
           1       0.86      0.96      0.91        25

    accuracy                           0.88        42
   macro avg       0.89      0.86      0.87        42
weighted avg       0.89      0.88      0.88        42



**La precisión total del modelo es de 0.88**

              precision    recall  f1-score   support

           0       0.93      0.76      0.84        17
           1       0.86      0.96      0.91        25

    accuracy                           0.88        42
    macro avg      0.89      0.86      0.87        42
    weighted avg   0.89      0.88      0.88        42